### PR TITLE
fix: compaction looping fixes

### PR DESCRIPTION
**Pull request by Joe-Blount**
This PR addresses issues found in another partition stuck compacting.  Two changes are made to address it getting stuck (which are the first two commits):
1)  When there is a lot of L1, and a lot more L0 coming, we now merge the L1 to L2 before we get all the L0s compacted.  This helps write amplification (as evidenced by the insta test updates for various big backlog scenarios), and it makes the highly backlogged scenarios easier to solve.  Since the L0s can overlap each other, and L1s can't, sometimes highly backlogged scenarios get painful with a lot of splitting and recompacting trying to rearrange the files so compactions can produce legal (non-overlapping) L1s.  Having many large L1s that overlap with the mess coming down from L0 adds to the challenge.  
2) Better/smarter checking for file grouping decisions to avoid setting us up to compact groups of files that undo previous splits

The catalog generated test case exposing this issue was around 4500 lines, so I (eventually) found a simpler test case that exercises a similar situation.

Note that the insta test updates for other scenarios show significant reductions in write amplification.  That's because the change in the first commit helps write amplification (3 level log structured merge should always cap the size of the intermediate level to optimize write amplification).

helps: https://github.com/influxdata/idpe/issues/17902

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: compactor/tests/layouts/backfill.rs [hunk 1]
```diff
@@ -954,56 +942,56 @@ async fn random_backfill_over_l2s() {
...
     - "**** 2 Output Files (parquet_file_id not yet assigned), 154mb total:"
     - "L2                                                                                                                 "
     - "L2.?[800,865] 1.05us 101mb|--------------------------L2.?---------------------------|                               "
     - "L2.?[866,899] 1.05us 53mb                                                           |-----------L2.?------------|  "
     - "Committing partition 1:"
-    - "  Soft Deleting 3 files: L2.9, L1.120, L1.124"
+    - "  Soft Deleting 3 files: L2.9, L1.113, L1.117"
     - "  Creating 2 files"
-    - "**** Final Output Files (4.58gb written)"
+    - "**** Final Output Files (4.06gb written)"
```
Comment (Joe-Blount) [inline compactor/tests/layouts/backfill.rs:975]: write amplification went down in previous test

File: compactor/tests/layouts/backfill.rs [hunk 2]
```diff
@@ -3020,63 +3008,66 @@ async fn actual_case_from_catalog_1() {
     - "WARNING: file L0.161[327,333] 336ns 183mb exceeds soft limit 100mb by more than 50%"
     - "WARNING: file L0.162[330,338] 340ns 231mb exceeds soft limit 100mb by more than 50%"
     - "WARNING: file L0.163[331,338] 341ns 232mb exceeds soft limit 100mb by more than 50%"
-    - "**** Final Output Files (17.64gb written)"
+    - "**** Final Output Files (15.47gb written)"
```
Comment (Joe-Blount) [inline compactor/tests/layouts/backfill.rs:3011]: write amplification went down in previous test

File: compactor/tests/layouts/many_files.rs [hunk 3]
```diff
@@ -4670,17 +4670,17 @@ async fn l0s_almost_needing_vertical_split() {
     - "L0.998[24,100] 1.02us    |-----------------------------------------L0.998-----------------------------------------|"
     - "L0.999[24,100] 1.02us    |-----------------------------------------L0.999-----------------------------------------|"
     - "L0.1000[24,100] 1.02us   |----------------------------------------L0.1000-----------------------------------------|"
-    - "**** Final Output Files (6.5gb written)"
+    - "**** Final Output Files (5.23gb written)"
```
Comment (Joe-Blount) [inline compactor/tests/layouts/many_files.rs:4673]: write amplification went down in previous test

File: compactor/tests/layouts/stuck.rs [hunk 4]
```diff
@@ -1111,81 +1111,78 @@ async fn stuck_l0() {
     - "L2.59[1686863759000000000,1686867839000000000] 1686928811.43s 96mb                                                              |--L2.59--|                 "
     - "L2.74[1686867899000000000,1686868319000000000] 1686928811.43s 14mb                                                                          |L2.74|         "
     - "L2.78[1686868379000000000,1686873599000000000] 1686928118.43s 39mb                                                                           |---L2.78----| "
-    - "**** Final Output Files (51.79gb written)"
-    - "L1                                                                                                                 "
-    - "L1.1998[1686873236374845980,1686873480877117666] 1686936871.55s 48mb                                                                                        |L1.1998|"
-    - "L1.1999[1686873480877117667,1686873599000000000] 1686936871.55s 23mb                                                                                         |L1.1999|"
+    - "**** Final Output Files (43.43gb written)"
```
Comment (Joe-Blount) [inline compactor/tests/layouts/stuck.rs:1114]: write amplification went down in previous test

File: compactor/tests/layouts/stuck.rs [hunk 5]
```diff
@@ -1702,31 +1699,32 @@ async fn stuck_l0_large_l0s() {
     - "L0.198[197,1970000] 197ns 10b|----------------------------------------L0.198-----------------------------------------| "
     - "L0.199[198,1980000] 198ns 10b|----------------------------------------L0.199-----------------------------------------| "
     - "L0.200[199,1990000] 199ns 10b|----------------------------------------L0.200-----------------------------------------| "
-    - "**** Final Output Files (14.48gb written)"
-    - "L1                                                                                                                 "
-    - "L1.3770[1752,1811] 199ns 62mb|L1.3770|                                                                                 "
+    - "**** Final Output Files (11.25gb written)"
```
Comment (Joe-Blount) [inline compactor/tests/layouts/stuck.rs:1702]: write amplification went down in previous test

File: compactor/tests/layouts/stuck.rs [hunk 6]
```diff
@@ -3027,17 +3025,805 @@ async fn split_precent_loop() {
     - "L1.3[1676005158277000000,1676010156669000000] 1676010160.05s 58mb                    |L1.3|                                                                "
     - "WARNING: file L0.40[1676020762355000000,1676036230752000000] 1676036233.84s 159mb exceeds soft limit 100mb by more than 50%"
     - "WARNING: file L0.43[1676039845773000000,1676063836202000000] 1676063839.07s 242mb exceeds soft limit 100mb by more than 50%"
-    - "**** Final Output Files (4.17gb written)"
+    - "**** Final Output Files (3.4gb written)"
```
Comment (Joe-Blount) [inline compactor/tests/layouts/stuck.rs:3028]: write amplification went down in previous test

File: compactor/src/components/round_info_source/mod.rs [hunk 7]
```diff
@@ -187,23 +191,53 @@ impl RoundInfoSource for LevelBasedRoundInfo {
...
+//
+// However, allowing intra-level overlaps in L0 but not L1/L2 adds extra challenge to compacting L0s to L1.
+// This is especially true when there are large quantitites of overlapping L0s and L1s, potentially resulting
+// in many split/compact cycles to resolve the overlaps.
+//
+// Since L1 & L2 only have inter-level overlaps, they can be compacted with just a few splits to align the L1s
+// with the L2s.  The relative ease of moving data from L1 to L2 provides additional motivation to compact the
+// L1s to L2s when a backlog of L0s exist. The easily solvable L1->L2 compaction can give us a clean slate in
+// L1, greatly simplifying the remaining L0->L1 compactions.
+fn get_start_level(files: &[ParquetFile], max_files: usize, max_bytes: usize) -> CompactionLevel {
```
Comment (wiedld) [inline compactor/src/components/round_info_source/mod.rs:208]: Confirmed the code does what it says, and this change does reduce write amplification.

In test case `very_big_overlapped_backlog()`:
* without this change: 68.33gb written
* with this change: 45.56gb written 🎉

File: compactor/src/components/split_or_compact/start_level_files_to_split.rs [hunk 8]
```diff
@@ -301,7 +301,26 @@ pub fn merge_small_l0_chains(
...
+                for f2 in merged_chains[prior_chain_idx as usize].iter() {
+                    if f.max_l0_created_at == f2.max_l0_created_at {
+                        matches += 1;
+                        break;
+                    }
+                }
+            }
+        }
+
+        // Merge it if: there a prior chain to merge with, and merging wouldn't make it too big, or undo a previous split
```
Comment (wiedld) [inline compactor/src/components/split_or_compact/start_level_files_to_split.rs:319]: Code does what it says.

This change alone reduces is responsible for:
* all of the write amplification reduction in test case `actual_case_from_catalog_1`
* all of the write amplification reduction in test case `l0s_almost_needing_vertical_split`
* the majority of the write amplification reduction in test case `stuck_l0`

**Review by wiedld — APPROVED** [review]
Nice.

**Pull request merged**